cmd/juju/machine/upgradeseries.go

```diff
@@ -305,30 +343,28 @@ func (c *upgradeSeriesCommand) handleUpgradeSeriesChange(ctx *cmd.Context, wid s
 	return nil
 }
 
-// UpgradeSeriesComplete completes a series for a given machine, that is
-// if a machine is marked as upgrading using UpgradeSeriesPrepare, then this
-// command will complete that process and the machine will no longer be marked
-// as upgrading.
+// UpgradeSeriesComplete completes a series upgrade for a given machine,
+// that is if a machine is marked as upgrading using UpgradeSeriesPrepare,
+// then this command will complete that process and the machine will no longer
+// be marked as upgrading.
 func (c *upgradeSeriesCommand) UpgradeSeriesComplete(ctx *cmd.Context) error {
-	var apiRoot api.Connection
-	var err error
-
-	if c.upgradeMachineSeriesClient == nil {
-		apiRoot, err = c.NewAPIRoot()
-		if err != nil {
-			return errors.Trace(err)
-		}
+	apiRoot, err := c.ensureAPIClients()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if apiRoot != nil {
 		defer apiRoot.Close()
-		c.upgradeMachineSeriesClient = machinemanager.NewClient(apiRoot)
 	}
 
-	err = c.upgradeMachineSeriesClient.UpgradeSeriesComplete(c.machineNumber)
-	if err != nil {
+	if err := c.upgradeMachineSeriesClient.UpgradeSeriesComplete(c.machineNumber); err != nil {
 		return errors.Trace(err)
 	}
 
-	err = c.handleNotifications(ctx)
-	if err != nil {
+	if err := c.handleNotifications(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := c.unpinLeaders(ctx); err != nil {
```

Comment:
Should there be a deferred here?

Reply:
Or perhaps even just a differed on somewhere in the `prepared` sub command to unpin in case of failure. I am assuming un/pinning is idempotent.